From: Egbert Eich <eich@suse.de>
Date: Sat Aug 30 09:57:29 2014 +0200
Subject: drm/panic: Add option to set panic mode
Patch-mainline: never

References: bnc#892114
Signed-off-by: Egbert Eich <eich@suse.com>

The option drm:panic allows to set the behavior of the fb helper
in panic mode:
 < 0 : default - restore when no reboot is scheduled
 = 0 : never restore
 > 0 : restore (unless immediate reboot is scheduled)

Signed-off-by: Egbert Eich <eich@suse.de>
---
 drivers/gpu/drm/drm_fb_helper.c | 15 ++++++++++++---
 1 file changed, 12 insertions(+), 3 deletions(-)

diff --git a/drivers/gpu/drm/drm_fb_helper.c b/drivers/gpu/drm/drm_fb_helper.c
index 1b207ee..03a1ebf 100644
--- a/drivers/gpu/drm/drm_fb_helper.c
+++ b/drivers/gpu/drm/drm_fb_helper.c
@@ -339,15 +339,24 @@ static bool drm_fb_helper_force_kernel_mode(void)
 	return error;
 }
 
+static int drm_fb_panic_mode = -1;
+module_param_named(panic, drm_fb_panic_mode, int, 0644);
+
 static int drm_fb_helper_panic(struct notifier_block *n, unsigned long ununsed,
 			void *panic_str)
 {
 	/*
 	 * It's a waste of time and effort to switch back to text console
-	 * if the kernel should reboot - since the fb_panic code may hang
-	 * it may even prevent rebooting.
+	 * if the kernel should reboot  before panic messages can be seen.
+	 * Since the fb_panic code may hang and may even prevent rebooting
+	 * we add an option to set the behavior:
+	 * drm:panic < 0 - default: don't switch when system is set to reboot.
+	 * drm:panic == 0: don't switch ever.
+	 * drm:panic > 0: switch when no immediate reboot is requested
+	 *                (original behavior)
 	 */
-	if (panic_timeout != 0)
+	if (drm_fb_panic_mode == 0 || panic_timeout < 0 ||
+	    (drm_fb_panic_mode < 0 && panic_timeout != 0))
 		return 0;
 
 	pr_err("panic occurred, switching back to text console\n");
